components/tidb_query/src/codec/mysql/duration.rs

```diff
@@ -244,217 +243,179 @@ mod parser {
-
-        duration.set_reserved(false);
-        Ok(duration)
+    #[inline]
+    pub fn to_secs(self) -> i64 {
+        self.nanos / NANOS_PER_SEC
     }
 
-    /// Returns the identity element of `Duration`
-    pub fn zero() -> Duration {
-        Duration(0)
+    /// Returns the number of seconds contained by this Duration as f64.
+    /// The returned value does include the fractional (nanosecond) part of the duration.
+    #[inline]
+    pub fn to_secs_f64(self) -> f64 {
+        self.nanos as f64 / NANOS_PER_SEC as f64
     }
 
-    /// Returns true if self is equal to the additive identity.
-    pub fn is_zero(mut self) -> bool {
-        self.set_neg(false);
-        self.set_fsp(0);
-        self.to_bits() == 0
+    #[inline]
+    pub fn to_millis(self) -> i64 {
+        self.nanos / NANOS_PER_MILLI
     }
 
-    /// Returns the absolute value of `Duration`
-    pub fn abs(mut self) -> Self {
-        self.set_neg(false);
-        self
+    #[inline]
+    pub fn to_micros(self) -> i64 {
+        self.nanos / NANOS_PER_MICRO
     }
 
-    /// Returns the fractional part of `Duration`, in whole microseconds.
-    pub fn subsec_micros(self) -> u32 {
-        self.micros()
+    #[inline]
+    pub fn to_nanos(self) -> i64 {
+        self.nanos
     }
 
-    /// Returns the number of whole seconds contained by this Duration.
-    pub fn to_secs(self) -> i32 {
-        let secs =
-            (self.hours() * SECS_PER_HOUR + self.minutes() * SECS_PER_MINUTE + self.secs()) as i32;
-
-        if self.get_neg() {
-            -secs
-        } else {
-            secs
+    /// Returns the identity element of `Duration`
+    #[inline]
+    pub fn zero() -> Duration {
+        Duration {
+            nanos: 0,
+            fsp: DEFAULT_FSP as u8,
         }
     }
 
-    /// Returns the number of seconds contained by this Duration as f64.
-    /// The returned value does include the fractional (nanosecond) part of the duration.
-    pub fn to_secs_f64(self) -> f64 {
-        let secs = f64::from(self.to_secs());
-        let micros = f64::from(self.subsec_micros()) * 1e-6;
-
-        secs + if self.get_neg() { -micros } else { micros }
+    /// Returns true if self is equal to the additive identity.
+    #[inline]
+    pub fn is_zero(self) -> bool {
+        self.nanos == 0
     }
 
-    /// Returns the `Duration` in whole nanoseconds
-    pub fn to_nanos(self) -> i64 {
-        let secs = i64::from(self.to_secs()) * NANOS_PER_SEC;
-        let micros = i64::from(self.subsec_micros());
+    /// Returns the absolute value of `Duration`
+    #[inline]
+    pub fn abs(self) -> Self {
+        Duration {
+            nanos: self.nanos.abs(),
+            ..self
+        }
+    }
 
-        secs + if self.get_neg() { -micros } else { micros } * 1000
+    pub fn from_secs(secs: i64, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
+        Ok(Duration {
+            nanos: secs.checked_mul(NANOS_PER_SEC).ok_or_else(|| {
```

Comment:
Do we need to check whether it overflows max duration?

Reply:
Done